In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

In [3]:
scores = {'activation':[],'loss':[],'accuracy':[]}

In [4]:
def create_model(activation):
    model = Sequential()
    model.add(Dense(128,input_shape=(784,), activation=activation))
    model.add(Dense(64, activation=activation))
    model.add(Dense(10, activation = 'softmax'))
    return model

# Part 1

## Problem 1

In [5]:
model = create_model('tanh')

In [6]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [7]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 1.0013 - accuracy: 0.7571
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 0.5051 - accuracy: 0.8741
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 0.4124 - accuracy: 0.8904
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3682 - accuracy: 0.8996
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3406 - accuracy: 0.9056
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3208 - accuracy: 0.9099
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3052 - accuracy: 0.9134
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2924 - accuracy: 0.9168
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2814 - accuracy: 0.9198
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2715 - accuracy: 0.9227

In [8]:
scores['activation'].append('tanh')
scores['loss'].append('cross entropy')
score = model.evaluate(X_test, Y_test, verbose=0)
scores['accuracy'].append(score[1])

In [9]:
print('Accuracy score for problem 1:',score[1])

Accuracy score for problem 1: 0.9399999976158142


## Problem 2

In [10]:
model = create_model('sigmoid')

In [11]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [12]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2862 - accuracy: 0.1563
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2279 - accuracy: 0.3243
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 2.1637 - accuracy: 0.5174
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 2.0737 - accuracy: 0.6042
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 1.9439 - accuracy: 0.6378
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 1.7691 - accuracy: 0.6667
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 1.5680 - accuracy: 0.6932
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 1.3749 - accuracy: 0.7217
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 1.2116 - accuracy: 0.7432
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 1.0807 - accuracy: 0.7633

In [13]:
scores['activation'].append('sigmoid')
scores['loss'].append('cross entropy')
score = model.evaluate(X_test, Y_test, verbose=0)
scores['accuracy'].append(score[1])

In [14]:
print('Accuracy score for problem 1:',score[1])

Accuracy score for problem 1: 0.859499990940094


## Problem 3

In [15]:
model = create_model('relu')

In [16]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [17]:
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 1.3225 - accuracy: 0.6617
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 0.5216 - accuracy: 0.8632
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3986 - accuracy: 0.8897
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3484 - accuracy: 0.9019
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3188 - accuracy: 0.9098
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2980 - accuracy: 0.9153
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2817 - accuracy: 0.9199
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2679 - accuracy: 0.9242
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2557 - accuracy: 0.9277
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2447 - accuracy: 0.9300

In [18]:
scores['activation'].append('relu')
scores['loss'].append('cross entropy')
score = model.evaluate(X_test, Y_test, verbose=0)
scores['accuracy'].append(score[1])

In [19]:
print('Accuracy score for problem 1:',score[1])

Accuracy score for problem 1: 0.9495000243186951


In [20]:
pd.DataFrame(scores)

activation           loss  accuracy
0       tanh  cross entropy    0.9400
1    sigmoid  cross entropy    0.8595
2       relu  cross entropy    0.9495

Relu performed best with an accuracy of 0.94

# Part 2